### ISSUE TITLE
Add possibility to use signal name as @let template variable

### ISSUE THREAD

**Issue by MikaStark** [angular/angular#56935]
### Description

It seams we cannot use the same name as a signal property like we does with `as` alias. This is a shame as it leads to the use of more verbose names like `@let userData = user()`.

The following code throw error `NG8016: Cannot read @let declaration 'user' before it has been defined.`

```html
@let user = user();
```

But this one works perfectly

```html
@if (user(); as user) {
  <!-- ... -->
}
```

### Proposed solution

```html
@let user = user()
```

or maybe simply infer signal extraction

```html
@let user;
```

### Alternatives considered

Any better idea ?

**Comment by crisbeto** [comment]
This works identically to `let` in TypeScript:
<img width="640" alt="image" src="https://github.com/angular/angular/assets/4450522/2800be1c-ea6a-4a13-b672-89d9b17ff94a">

**Comment by MikaStark** [comment]
Yes but it decreases developer experience with signals comparing to « as » alias

**Comment by crisbeto** [comment]
This will lead to some weird cases like this:

```
{{data()}} <!-- Signal coming from the component instance -->

@let data = data() + 'foo';

{{data}} <!-- Local template variable -->
```

**Comment by MikaStark** [comment]
This have been the case with `as` alias since the beginning (`*ngIf`) and it was pretty okay, isn't it ?

**Comment by MikaStark** [comment]
I understand your point @crisbeto but in fact this is something Angular devs are used to (or at least me)

https://stackblitz.com/edit/stackblitz-starters-agtvef?file=src%2Fmain.ts

**Comment by crisbeto** [comment]
The semantics here are a bit different though. With `@if`, it represents something like this:

```
if (data()) {
  const data = data()!;
}
```

Whereas `@let` is meant to be a `let data = data();`.

**Comment by MikaStark** [comment]
Here's a revised version of your GitHub comment to make it gentler and more polished:

---

Thank you for pointing that out. You are absolutely right. The goal of this issue is to simplify code writing. The `as` alias seems a bit conceptually strange since it doesn't exist in the broader JS ecosystem.

However, I believe `@let` is a great opportunity to write cleaner code. To be fair, the kind of property extraction that allows the `as` alias is indeed very useful for developers using signals. Until recently, we had to use workarounds to handle signal type checking in templates, such as:

```html
@if (data(); as data) {
  {{ data.label }}
}

<!-- or even -->

@if (data()) {
  {{ data()!.label }}
}
```

Now, with `@let`, we need to adjust our approach slightly. While it's not a huge change, renaming a variable just for the template can feel overwhelming. It's already challenging to come up with good variable names, and having to do it twice adds to the complexity.

```html
@let someData = data(); <!-- challenge: find a better name than "someData" -->

@if (someData) {
  {{ someData.label }}
}
```

Maybe my proposal is not the good one. but I'm sure we can find a better way.

**Comment by JeanMeche** [comment]
Signal narrowing is a separate issue and is not intended to be fixed by this. See #49161, to track that specific issue.

**Comment by MikaStark** [comment]
You're right @JeanMeche, but maybe this feature request can address it too, don't you think ?

**Comment by JeanMeche** [comment]
This is not ideal. 
Narrowing is purely a typing issue, it should not be fixed with runtime change. 

Here is a simple demo that shows the implications, in terms of generated instructions : https://jeanmeche.github.io/angular-compiler-output/?template=PGRpdj4KICBAbGV0IG15VmFyID0gNTvEhSB7e8SNxI9yfX0KPC%2FEgcSD

**Comment by MikaStark** [comment]
@JeanMeche you convinced me.

**Comment by crisbeto** [comment]
For what it's worth, you should be able to write your example like this if you want to keep the `data` name:

```
@let data = this.data();
```

**Comment by MikaStark** [comment]
Thank you. That's a solution indeed. But just like Matthieu guess, my real concern was, in fact, type narrowing and so it wouldn't be a nice approach.

Anyway, thank you for your time and patience :)

**Comment by JeanMeche** [comment]
For anyone reading this issue, be wary of suggested `@let data = this.data();` pattern. 

In v18, this will shadow the `data` attribute. 
This is a known issue that will be fixed with #55183. But because of the breaking nature of this change, it will be delayed until v19.
